src/System.Linq.Expressions/src/System/Linq/Expressions/LambdaExpression.cs

```diff
@@ -18,36 +19,22 @@ namespace System.Linq.Expressions
     /// Lambda expressions take input through parameters and are expected to be fully bound.
     /// </remarks>
     [DebuggerTypeProxy(typeof(LambdaExpressionProxy))]
-    public abstract class LambdaExpression : Expression
+    public abstract class LambdaExpression : Expression, IParameterProvider
     {
-        private readonly string _name;
         private readonly Expression _body;
-        private readonly ReadOnlyCollection<ParameterExpression> _parameters;
-        private readonly Type _delegateType;
-        private readonly bool _tailCall;
 
-        internal LambdaExpression(
-            Type delegateType,
-            string name,
-            Expression body,
-            bool tailCall,
-            ReadOnlyCollection<ParameterExpression> parameters
-        )
+        internal LambdaExpression(Expression body)
         {
-            Debug.Assert(delegateType != null);
-
-            _name = name;
             _body = body;
-            _parameters = parameters;
-            _delegateType = delegateType;
-            _tailCall = tailCall;
         }
 
         /// <summary>
         /// Gets the static type of the expression that this <see cref="Expression"/> represents. (Inherited from <see cref="Expression"/>.)
         /// </summary>
-        /// <returns>The <see cref="System.Type"/> that represents the static type of the expression.</returns>
-        public sealed override Type Type => _delegateType;
+        /// <returns>The <see cref="Type"/> that represents the static type of the expression.</returns>
+        public sealed override Type Type => TypeCore;
+
+        protected abstract Type TypeCore { get; }
```

Comment:
Being `protected` makes this part of the external-visible interface. Although nothing can actually derive from the type because its constructor is `internal`, this should probably be `internal` too.